File: OsmAnd/src/net/osmand/plus/settings/enums/LengthUnits.java
package net.osmand.plus.settings.enums;

import static net.osmand.plus.utils.OsmAndFormatter.CENTIMETERS_IN_ONE_METER;
import static net.osmand.plus.utils.OsmAndFormatter.FEET_IN_ONE_METER;
import static net.osmand.plus.utils.OsmAndFormatter.INCHES_IN_ONE_METER;
import static net.osmand.plus.utils.OsmAndFormatter.YARDS_IN_ONE_METER;

import androidx.annotation.StringRes;

import net.osmand.plus.R;
import net.osmand.util.CollectionUtils;

public enum LengthUnits implements MeasurementUnits {

	METERS(R.string.shared_string_meters, R.string.m, 1),
	CENTIMETERS(R.string.shared_string_centimeters, R.string.centimeter, CENTIMETERS_IN_ONE_METER),
	INCHES(R.string.shared_string_inches, R.string.inch, INCHES_IN_ONE_METER),
	FEET(R.string.shared_string_feet, R.string.foot, FEET_IN_ONE_METER),
	YARDS(R.string.shared_string_yards, R.string.yard, YARDS_IN_ONE_METER);

	private final int symbolResId;
	private final int nameResId;
	private final float coefficient;

	LengthUnits(@StringRes int nameResId, @StringRes int symbolResId, float coefficient) {
		this.nameResId = nameResId;
		this.symbolResId = symbolResId;
		this.coefficient = coefficient;
	}

	@StringRes
	@Override
	public int getSymbolResId() {
		return symbolResId;
	}

	@StringRes
	@Override
	public int getNameResId() {
		return nameResId;
	}

	@Override
	public float getConversionCoefficient() {
		return coefficient;
	}

	@Override
	public boolean isMetric() {
		return CollectionUtils.equalsToAny(this, METERS, CENTIMETERS);
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/units/LengthUnits.kt
package net.osmand.shared.units

object LengthConstants {
	const val CENTIMETERS_IN_METER = 100.0
	const val INCHES_IN_METER = 39.37007874015748
	const val FEET_IN_METER = 3.2808398950131235
	const val YARDS_IN_METER = 1.0936132983377078
	const val KILOMETERS_IN_METER = 0.001
	const val MILES_IN_METER = 0.0006213711922373339
	const val NAUTICAL_MILES_IN_METER = 0.0005399568034557235
}

enum class LengthUnits(
	override val nameKey: String,
	override val symbolKey: String,
	override val conversionCoefficient: Double,
	override val system: MeasurementSystem
) : MeasurementUnit<LengthUnits> {

	METERS(
		nameKey = "shared_string_meters",
		symbolKey = "m",
		conversionCoefficient = 1.0,
		system = MeasurementSystem.METRIC
	),

	KILOMETERS(
		nameKey = "shared_string_kilometers",
		symbolKey = "km",
		conversionCoefficient = LengthConstants.KILOMETERS_IN_METER,
		system = MeasurementSystem.METRIC
	),

	CENTIMETERS(
		nameKey = "shared_string_centimeters",
		symbolKey = "centimeter",
		conversionCoefficient = LengthConstants.CENTIMETERS_IN_METER,
		system = MeasurementSystem.METRIC
	),

	MILES(
		nameKey = "miles",
		symbolKey = "mile",
		conversionCoefficient = LengthConstants.MILES_IN_METER,
		system = MeasurementSystem.IMPERIAL
	),

	NAUTICAL_MILES(
		nameKey = "si_nm",
		symbolKey = "nm",
		conversionCoefficient = LengthConstants.NAUTICAL_MILES_IN_METER,
		system = MeasurementSystem.IMPERIAL
	),

	INCHES(
		nameKey = "shared_string_inches",
		symbolKey = "inch",
		conversionCoefficient = LengthConstants.INCHES_IN_METER,
		system = MeasurementSystem.IMPERIAL
	),

	FEET(
		nameKey = "shared_string_feet",
		symbolKey = "foot",
		conversionCoefficient = LengthConstants.FEET_IN_METER,
		system = MeasurementSystem.IMPERIAL
	),

	YARDS(
		nameKey = "shared_string_yards",
		symbolKey = "yard",
		conversionCoefficient = LengthConstants.YARDS_IN_METER,
		system = MeasurementSystem.IMPERIAL
	);
}